### ISSUE TITLE
Integrating Brunch with Create-React-App

### ISSUE THREAD

**Issue by adrianmc** [brunch/brunch#1466]
# TL;DR

Dan Abramov (@gaearon) has [hinted](https://twitter.com/dan_abramov/status/760439982595203072) that he might like to see a Brunch version of [Create-React-App (CRA)](https://github.com/facebookincubator/create-react-app). I think doing so will help both the React and Brunch communities tremendously. It would be nice if we can get some contributors working on this together.

#### About Create-React-App

[Official Facebook Blogpost Announcing Create-React-App](https://facebook.github.io/react/blog/2016/07/22/create-apps-with-no-configuration.html)
[Github for Create-React-App](https://github.com/facebookincubator/create-react-app)

#### Work in Progress
[My Working Branch (super hacky at the moment) — PRs Welcome!](https://github.com/adrianmc/create-react-app/tree/brunch-experimental)

#### Using
[Brunch React Skeleton](https://github.com/brunch/with-react)

## Description

This isn't so much of a bug as it is an opportunity to get more exposure for the Brunch build tool, which I'm sure we'll all agree is one of the best and fastest ways to be productive with React. With Brunch, we can achieve simplicity and speed while still maintaining a lot of configurability.

I recently updated my article on Brunch vs Webpack [here](https://medium.com/@adrianli/the-zero-config-javascript-build-tool-1914d71d749c). And I tweeted it out to the twittersphere. I was very surprised when I received the following tweet from none other than Dan Abramov himself:

![screen shot 2016-08-02 at 10 23 51 pm](https://cloud.githubusercontent.com/assets/943555/17332027/e82106dc-58ff-11e6-9e27-829a95fce3f5.png)

— [Link to Twitter Conversation](https://twitter.com/dan_abramov/status/760439982595203072)

## Current CRA behavior

I'll briefly explain what CRA does here, but I highly recommend you actually go try to run it yourself or at least read the [blog article](https://facebook.github.io/react/blog/2016/07/22/create-apps-with-no-configuration.html) announcing it.

The main functions of CRA are: `start`, `build`, `eject`.

Since CRA is a command-line tool, it works by forwarding calls to a [scripts folder](https://github.com/facebookincubator/create-react-app/tree/master/scripts) that contains node scripts written in javascript. [This file](https://github.com/facebookincubator/create-react-app/blob/master/global-cli/index.js) manages the forwarding of the command line arguments to the scripts folder.

- [`start`](https://github.com/facebookincubator/create-react-app/blob/master/scripts/start.js): This [creates](https://github.com/facebookincubator/create-react-app/blob/master/scripts/start.js#L72-L130) a Webpack compiler using a config file and then [starts](https://github.com/facebookincubator/create-react-app/blob/master/scripts/start.js#L152-L171) a WebpackDevServer. 
- [`build`](https://github.com/facebookincubator/create-react-app/blob/master/scripts/build.js): This simply uses Webpack to compile an optimized production build.
- [`eject`](https://github.com/facebookincubator/create-react-app/blob/master/scripts/eject.js): This converts the simplified app into a complete Webpack app with all of its numerous config files.

I should also mention that [`init.js`](https://github.com/facebookincubator/create-react-app/blob/master/scripts/init.js) is how the simplified app is first created. Usually, the user would run the following command to create their app:

```
create-react-app my-app
```

## What I have done

On my [working branch](https://github.com/adrianmc/create-react-app/tree/brunch-experimental), I have already (kind of) gotten the `start` functionality to work with the [React Skeleton repo](https://github.com/brunch/with-react). It's a little hacky though, because I simply use [shelljs](https://github.com/shelljs/shelljs) to run `brunch watch --server`. You'll notice that I even had to `cd` into the `template` folder for development purposes. Obviously, this should be taken out later on.

Note also, that I [changed](https://github.com/adrianmc/create-react-app/commit/70976040b508e76ef341a5dd7a696aaa01f0faa1) the build output folder to `/build` instead of the default `/public`.

Anyway, go to the [branch](https://github.com/adrianmc/create-react-app/tree/brunch-experimental) and take a look at my [commits](https://github.com/adrianmc/create-react-app/commits/brunch-experimental) to see what I have done (which unfortunately, is not too much yet).

## What needs to be done

I'll only talk about the `start` functionality as the `build` and `eject` functionality has yet to be attempted.

### Flatten folder structure

The output from CRA is currently a single `src` folder inside their app folder. And inside this `src` folder, they have all of the styles and components in a flat hierarchy. In contrast, our [React Skeleton](https://github.com/brunch/with-react/tree/master/app) has 3 folders: `assets`, `components`, and `styles`. I suspect Dan will say that having these folders is unnecessarily confusing.

**CRA:** You can see this from their template folder [here](https://github.com/facebookincubator/create-react-app/tree/master/template).
**Brunch:** And compare it with the React Skeleton folder [here](https://github.com/brunch/with-react/tree/master/app).

### Move `index.html` to the root folder

Currently, our `index.html` resides in the `assets` folder because, by default, Brunch treats anything inside an `assets` folder as a static asset and will simply copy it over to the build folder. 

Unfortunately, it seems that the preferred approach with CRA is to have the `index.html` at the root of the project folder. This is annoying because this means we would have to add to our configuration file to take care of this.

### Launch Brunch with a config file elsewhere

I'm not certain if this is possible, but one of the big advantages of CRA is that it abstracts away all config files. When I launch Brunch, however, it always requires a `brunch-config.js` in the root of the project.

I was wondering if there was a way to run Brunch while pointing to a specific config file instead. This would make the generated app much simpler. And when we need to implement the `eject` functionality, we can simply move the config file back into the root of the folder.

Perhaps something like this:

```
> brunch watch --server --config <path-to-config-file>
```

**EDIT:** I did some more digging regarding specifying a config file. Unfortunately, it seems to be a deprecated feature. Scroll down for more details.

**Comment by paulmillr** [comment]
Totally 👍

**Comment by paulmillr** [comment]
> but one of the big advantages of CRA is that it abstracts away all config files

The "abstraction" is simply moving the file into a subdirectory if i'm not mistaken. Not a "huge" abstraction. Nevertheless if they want to standardize it i'm all for it.

**Comment by adrianmc** [comment]
@paulmillr yeah, I don't think it's anything too difficult. As a Brunch member, do you have any recommendations on how we should approach this project together?

**Comment by colinbate** [comment]
FWIW, when you use Brunch, you almost don't need something like CRA since brunch and a good skeleton are really all you need.

**Comment by adrianmc** [comment]
@colinbate I absolutely agree. Honestly, Brunch is just SO much simpler that I wish that they'd recommend Brunch instead of having built CRA. Nevertheless, I want to do this as an exercise and for promoting Brunch.

Only problem is, I'm not sure how you can run brunch while pointing to a config file in another directory. It certainly does not look documented, but let me poke around the source code to see if it's already possible.

**Comment by adrianmc** [comment]
Apparently Brunch used to have the ability to specify a specific config file. It says so in the guide [here](https://github.com/brunch/brunch-guide/blob/master/content/en/chapter03-conventions-and-defaults.md#configuration-files):

> Customizing: a CLI option used to let you specify a config file, but that is deprecated in favor of a single file with per-environment overrides, something we'll illustrate in a later chapter.

But since it is deprecated, the feature is no longer available. If you try to use it, [this bit of code](https://github.com/brunch/brunch/blob/master/lib/commands.js#L51-L54) will trigger this error:

```js
// Deprecations
if (hasArg(['-c', '--config'])) {
  return '--config has been removed. Use `-e / --environment` and specify custom envorinments in config';
}
```

Therefore, in order to point to a config file outside of the folder, we'd need to actually change Brunch. Seeing as this is not really feasible in the short term, we might have to do something hacky.

It's possible to temporarily copy the config file into the folder immediately before launching the brunch server. However, that's a little TOO hacky, even for my taste. It brings up issues like: how are we going to get rid of the config file if the server is suddenly stopped?

**Comment by es128** [comment]
Any advantage other than just the ability to hide it in a folder?

What if `config/brunch.js` was simply added as another alternative path to look for the config in?

**Comment by paulmillr** [comment]
@es128 yep

**Comment by adrianmc** [comment]
@es128 I really wanted to replicate CRA exactly, but that's a fair and worthwhile compromise.

**Comment by colinbate** [comment]
In the CRA world, they don't have a `config` folder, the config is all part of the single `devDependency` that they install. I'm not sure that brunch would even work with that setup given its use of the `package.json` as a form of of configuration.

@paulmillr Is it possible to specify the brunch plugins without putting them in the `package.json`? I tried setting them "on" in the config, but that doesn't work if they aren't known already.

**Comment by paulmillr** [comment]
@colinbate not possible. where would we store them? They need to physically present somewhere.

**Comment by colinbate** [comment]
@paulmillr They'd be installed transitively via another package. I guess you'd either need to enumerate them explicitly in the config file, or explore the `node_modules` folder for them.

**Comment by colinbate** [comment]
Honestly I'm not suggesting that we change Brunch to mirror exactly what CRA does because as I see it, Brunch is an alternative to CRA, not a tool that can be shoehorned into it. Particularly since Brunch is also a project generator as well as a build tool.

@adrianmc I'd say if you want to offer up a Brunch flavoured alternative to CRA, then create the appropriate skeleton using their opinionated project template and add a brunch-config to work with the breaks in conventions. I think that demonstrating that would be a better testament to the power of Brunch. It doesn't need a suite of scripts to get it to work.

**Comment by colinbate** [comment]
And if desired, they could still have `create-react-app <proj>` as a wrapper around something like `brunch new -s facebook\react-skeleton <proj>`

**Comment by colinbate** [comment]
One thing that Brunch could add which is nice, is to poke the current project/folder name into the `package.json` file it gets from the skeleton.

**Comment by adrianmc** [comment]
> If you want to offer up a Brunch flavoured alternative to CRA, then create the appropriate skeleton using their opinionated project template and add a brunch-config to work with the breaks in conventions.

@colinbate I like this direction. I'll give it some more thought, but feel free to contribute as well.

**Comment by adrianmc** [comment]
For anyone still following, I'm starting the CLI tool on this repo: https://github.com/adrianmc/create-react-app-brunch

**Comment by paulmillr** [comment]
Some updates on the work i've done:

1. The skeleton that works already without `package.json` deps and brunch config: https://github.com/brunch/cra-brunch-test
2. The mediator that made the skeleton possible; alternative to `react-scripts`: https://github.com/brunch/cra-brunch

**Comment by paulmillr** [comment]
=> gh-1489
